packages/node/src/integrations/anr/worker.ts

```diff
@@ -0,0 +1,210 @@
+import {
+  createEventEnvelope,
+  createSessionEnvelope,
+  getEnvelopeEndpointWithUrlEncodedAuth,
+  makeSession,
+  updateSession,
+} from '@sentry/core';
+import type { Event, Session, StackFrame, TraceContext } from '@sentry/types';
+import { callFrameToStackFrame, stripSentryFramesAndReverse, watchdogTimer } from '@sentry/utils';
+import { Session as InspectorSession } from 'inspector';
+import { parentPort, workerData } from 'worker_threads';
+import { makeNodeTransport } from '../../transports';
+import type { WorkerStartData } from './common';
+
+type VoidFunction = () => void;
+type InspectorSessionNodeV12 = InspectorSession & { connectToMainThread: VoidFunction };
+
+const options: WorkerStartData = workerData;
+let session: Session | undefined;
+let hasSentAnrEvent = false;
+
+function log(msg: string): void {
+  if (options.debug) {
+    // eslint-disable-next-line no-console
+    console.log(`[ANR Worker] ${msg}`);
+  }
+}
+
+const url = getEnvelopeEndpointWithUrlEncodedAuth(options.dsn);
+const transport = makeNodeTransport({
+  url,
+  recordDroppedEvent: () => {
+    //
+  },
+});
+
+async function sendAbnormalSession(): Promise<void> {
+  // of we have an existing session passed from the main thread, send it as abnormal
+  if (session) {
+    log('Sending abnormal session');
+    updateSession(session, { status: 'abnormal', abnormal_mechanism: 'anr_foreground' });
+
+    log(JSON.stringify(session));
+
+    const envelope = createSessionEnvelope(session, options.dsn, options.sdkMetadata);
+    await transport.send(envelope);
+
+    try {
+      // Notify the main process that the session has ended so the session can be cleared from the scope
+      parentPort?.postMessage('session-ended');
+    } catch (_) {
+      // ignore
+    }
+  }
+}
+
+log('Started');
+
+async function sendAnrEvent(frames?: StackFrame[], traceContext?: TraceContext): Promise<void> {
+  if (hasSentAnrEvent) {
+    return;
+  }
+
+  hasSentAnrEvent = true;
+
+  await sendAbnormalSession();
+
+  log('Sending event');
+
+  const event: Event = {
+    sdk: options.sdkMetadata.sdk,
+    contexts: { ...options.contexts, trace: traceContext },
+    release: options.release,
+    environment: options.environment,
+    dist: options.dist,
+    platform: 'node',
+    level: 'error',
+    exception: {
+      values: [
+        {
+          type: 'ApplicationNotResponding',
+          value: `Application Not Responding for at least ${options.anrThreshold} ms`,
+          stacktrace: { frames },
+          // This ensures the UI doesn't say 'Crashed in' for the stack trace
+          mechanism: { type: 'ANR' },
+        },
+      ],
+    },
+    tags: { 'process.name': 'ANR' },
+  };
+
+  log(JSON.stringify(event));
```

Comment:
should we add additional message to this log?

Reply:
This is almost entirely there to make the worker thread easier to test and debug:
https://github.com/getsentry/sentry-javascript/pull/9823/files#diff-0af0ef940798370fb6cdd7d03fa822ab1e38d9b4cc9c50231f1210ad8cd99b3fL10-L25

There is logging surrounding these JSON objects that give some context:

<img width="401" alt="image" src="https://github.com/getsentry/sentry-javascript/assets/1150298/bb81160c-5b7a-4dbc-b43c-6d95d92bb198">